doc/wavetable-communication.md

```diff
@@ -0,0 +1,184 @@
+# Wavetable communication
+
+This explains how RT and non-RT threads communicate when new waves for the
+wavetable need to be generated.
+
+## Wavetable layout
+
+A Tensor3 (3D-Tensor) has the following contents:
+
+```
+  ooooooo 1D-Tensor (frequeny values in Hz)      --
+                                                  |
+  ooooooo 1D-Tensor (semantic values, e.g.        | "wavetable scales"
+                     random seeds or              |
+                     basewave function params)   --
+
+  ooooooo 1D-Tensor (float buffer = 1 wave)      --
+  ^                                               |
+  |                                               | "wavetable data"
+  o...ooo 2D Tensor (Tensor1s for each freq)      | or just
+  ^       with ringbuffer semantics (see below)   | "waves"
+  |       (MW produces, AD note consumes)         |
+  |                                               |
+  o...ooo 3D Tensor (Tensor2s for each semantic) --
+
+```
+
+Ringbuffer layout:
+
+```
+  no required space holder (r=w is unambiguous internally)             wrap
+         |                                                               |
+         v               |--reserved for write--|                        v
+  ...----|--can be read--|                      |--can be reserved next...
+  ...oooooooo............oooo...................oooo......................
+         |               |                      |
+         r               w_delayed              w              <- pointers
+```
+
+Wavetable layout:
+
+```
+ o scales for ringbuffer (Tensor1, semantics + freqs)
+ o ringbuffer (Tensor3, wave data)
+
+ o timestamp of params for current ringbuffer
+ o timestamp of params for next ("to be filled") ringbuffer
+
+ o wavetable mode (e.g. random seeds, basefunc parameter, ...)
+```
+
+ADnote contains two wavetables per OscilGen:
+
+- current wavetable (in used, can be refilled with new waves)
+- next wavetable (under construction after parameter change)
+
```

Comment:
This feels like we should have a unit test that verifies nothing 'fun' happens with:

- Start notes
- Change wavetable
- Start notes
- Change wavetable
- Start notes
- Stop notes

Reply:
Reading add synth, this seems to only have risks in the wavetable modulation case as otherwise the wavetable is used to initialize add synth sample buffers similar to the existing git HEAD.

Reply:
In non-wavetable mode, the formulation "the wavetable is used to initialize add synth sample buffers similar to the existing git HEAD" is a bit difficult, see this code from ADnote.cpp:

```c++
            for(int k = 0; k < vce.unison_size; ++k)
            {
                if(waveTables)
                {
                    assert(pars.VoicePar[nvoice].tableMod); // did you allocate ADnoteParameters and not assign its "table" member?
                    const float* bufferInTable = pars.VoicePar[nvoice].tableMod->getNextWaveToConsume(tmp).data();
                    std::copy(bufferInTable, bufferInTable + synth.oscilsize, vce.FMSmp);
                    vce.oscposhiFM[k] = pars.VoicePar[nvoice].FmGn->getFinalOutpos() % synth.oscilsize;
                }
                else // debug code, testing only
                {
                    vce.oscposhiFM[k] = (vce.oscposhi[k]
                                             + pars.VoicePar[vc].FmGn->get(
                                                 vce.FMSmp, tmp))
                                            % synth.oscilsize;
                }
            }
```

First of all, the `if(waveTables)` is only ever active in tests, in the main executable, it is always false. In the main program (not the tests), the behavior is very different to the git HEAD because we do not call `OscilGen::get()` at all (but probably you got this and meant something different), but only take data from the wavetable.

Also, the swapping occurs both in non-wavetable and wavetable mode, at the time of a parameter change which forces regeneration of the whole table. The only difference is that additionally, in non-wavetable mode, RT can request single waves to be updated (because they got used) instead of the whole table.

In any case, the scenario you described does make sense to test. Until now, I was spare with tests because I did not want to spend the effort for a feature which might not make it. However, such tests are indeed planned.

Reply:
Sounds good to me. I guess I misunderstood the boolean as I initially read it as "The voice is operating in a mode where it copies samples out of the wavetable" vs "The voice is operating in a mode where it uses the wavetable data directly."